userdocs/diagnostics/trailing-closure-matching.md

```diff
@@ -0,0 +1,99 @@
+# Argument Matching for Trailing Closures
+
+In Swift, calling a function with one or more trailing closure arguments requires the label of the first trailing closure argument to be omitted. As a result, the compiler must consider additional context when determining which function parameter the trailing closure should satisfy.
+
+Before Swift 5.3, the compiler used a backward scanning rule to match a trailing closure to a function parameter. Starting from the end of the parameter list, it moved backwards until finding a parameter compatible with trailing closures. This could sometimes lead to unexpected behavior. Consider the following example:
+
+```swift
+func animate(
+  withDuration duration: Double, 
+  animations: () -> Void, 
+  completion: (() -> Void)? = nil
+) {}
+
+// OK
+animate(withDuration: 0.3, animations: { /* Animate Something */ }) {
+  // Done!
+}
+
+// error: missing argument for parameter 'animations' in call
+animate(withDuration: 0.3) {
+  // Animate Something
+}
+```
+
+The second call to `animate` results in a compiler error because the backward scanning rule matches the trailing closure to the `completion` parameter instead of `animations`.
+
+Beginning in Swift 5.3, the compiler uses a new, forward scanning rule to match trailing closures to function parameters. When matching function arguments to parameters, the forward scan first matches all non-trailing arguments from left-to-right. Then, it continues matching trailing closures in a left-to-right manner. This leads to more predictable and easy-to-understand behvavior in many situations. With the new rule, the example above now works as expected without any modifications:
+
+```swift
+// Remains valid
+animate(withDuration: 0.3, animations: { /* Animate Something */ }) {
+  // Done!
+}
+
+// Also OK!
+animate(withDuration: 0.3) {
+  // Animate Something
+}
+```
+
+When scanning forwards to match an unlabeled trailing closure argument, the compiler may sometimes need to "skip over" defaulted and variadic arguments. The new rule will skip any parameter that does not structurally resemble a function type. This allows writing a modified version of the above example where `withDuration` also has a default argument value:
+
+```swift
+func animate(
+  withDuration duration: Double = 1.0, 
+  animations: () -> Void, 
+  completion: (() -> Void)? = nil
+) {}
+
+// Allowed! The forward scanning rule skips `withDuration` because it does not structurally resemble a function type.
```

Comment:
Wrap to 80 characters?

```suggestion
// Allowed! The forward scanning rule skips `withDuration` because it does not
// structurally resemble a function type.
```